# Heroes of Pymoli Homework

![Fantasy](Images/Fantasy.png)

Congratulations! After a lot of hard work in the data munging mines, you've landed a job as Lead Analyst for an independent gaming company. You've been assigned the task of analyzing the data for their most recent fantasy game Heroes of Pymoli.

Like many others in its genre, the game is free-to-play, but players are encouraged to purchase optional items that enhance their playing experience. As a first task, the company would like you to generate a report that breaks down the game's purchasing data into meaningful insights.

## How to read this homework
You will find that each task has it's instruction before the code to solve it, at the end and before a new task the solution can be found with a visual data frame representation.

At the end of this note book you'll find the observable trends with some insights.

In [1]:
# Dependencies imports
import pandas as pd

In [2]:
# Checking csv for successfull read
csv_path = 'Resources/purchase_data.csv'
purchase_data = pd.read_csv(csv_path)
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

### Player Count

* Total Number of Players
---------------------------------

In [3]:
# Number of players
number_of_players = len(purchase_data['SN'].value_counts())
number_of_players
players_df = pd.DataFrame({
  'Total Players': [number_of_players]  
})
players_df

Total Players
0            576

### Result Purchasing Analysis (Total)

* Number of Unique Items
* Average Purchase Price
* Total Number of Purchases
* Total Revenue
-----------------------------

In [4]:
# Number of unique items
number_of_unique_items = len(purchase_data['Item Name'].unique())
number_of_unique_items

179

In [5]:
# Average purchase price
avg_price = round(purchase_data['Price'].mean(),2)
avg_price

3.05

In [6]:
# Total number of purchases
total_purchases = purchase_data['Purchase ID'].count()
total_purchases

780

In [7]:
# Total revenue 
total_revenue = purchase_data['Price'].sum()
total_revenue

2379.77

In [8]:
pruchase_analysis_total_df = pd.DataFrame({
    'Number of Unique Items': [number_of_unique_items],
    'Average Price': [avg_price],
    'Number of Purchases': [total_purchases],
    'Total Revenue': [total_revenue]
})
pruchase_analysis_total_df['Average Price'] = pruchase_analysis_total_df['Average Price'].map('${:.2f}'.format)
pruchase_analysis_total_df['Total Revenue'] = pruchase_analysis_total_df['Total Revenue'].map('${:.2f}'.format)
pruchase_analysis_total_df

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780      $2379.77

### Result Gender Demographics

* Percentage and Count of Male Players
* Percentage and Count of Female Players
* Percentage and Count of Other / Non-Disclosed
-----------

In [9]:
# Percentage and Count of Male Players
male_players = len(purchase_data.loc[purchase_data['Gender'] == 'Male']['SN'].value_counts())
male_players

484

In [10]:
male_pct = round((male_players/number_of_players)*100, 2)
male_pct

84.03

In [11]:
# Percentage and Count of Female Players
female_players = len(purchase_data.loc[purchase_data['Gender'] == 'Female']['SN'].value_counts())
female_players

81

In [12]:
female_pct = round((female_players/number_of_players)*100, 2)
female_pct

14.06

In [13]:
# Percentage and Count of Other/Non-Disclosed
other_players = len(purchase_data.loc[purchase_data['Gender'] == 'Other / Non-Disclosed']['SN'].value_counts())
other_players

11

In [14]:
other_pct = round((other_players/number_of_players)*100, 2)
other_pct

1.91

In [15]:
gender_demo_df = pd.DataFrame([
    {
        'Gender': 'Male',
        'Total Count': male_players,
        'Percentage of Players': male_pct
    },
    {
        'Gender': 'Feale',
        'Total Count': female_players,
        'Percentage of Players': female_pct
    },
    {
        'Gender': 'Others/Non-Disclosed',
        'Total Count': other_players,
        'Percentage of Players': other_pct
    },
])
gender_demo_df = gender_demo_df.set_index('Gender')
gender_demo_df['Percentage of Players'] = gender_demo_df['Percentage of Players'].map('{:.2f}%'.format)
gender_demo_df.index.name = None
gender_demo_df

Total Count Percentage of Players
Male                          484                84.03%
Feale                          81                14.06%
Others/Non-Disclosed           11                 1.91%

### Result Purchasing Analysis (Gender)

* The below each broken by gender
  * Purchase Count
  * Average Purchase Price
  * Total Purchase Value
  * Average Purchase Total per Person by Gender
-------------------

In [16]:
# New male purchase data frame
male_purchase_data = purchase_data.loc[purchase_data['Gender'] == 'Male']
male_purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [17]:
# Male purchase count
male_purchase = len(male_purchase_data)
male_purchase

652

In [18]:
# Average male purchase price
male_mean_price = round(male_purchase_data['Price'].mean(),2)
male_mean_price

3.02

In [19]:
# Total male purchase 
male_total_purchase = round(male_purchase_data['Price'].sum(),2)
male_total_purchase

1967.64

In [20]:
# Average purchase total per person by males
male_avg_purchase = round(male_total_purchase/male_players,2)
male_avg_purchase

4.07

In [21]:
# New female purchase data frame
female_purchase_data = purchase_data.loc[purchase_data['Gender'] == 'Female']
female_purchase_data.head()

Purchase ID         SN  Age  Gender  Item ID  \
15           15  Lisassa64   21  Female       98   
18           18  Reunasu60   22  Female       82   
38           38   Reulae52   10  Female      116   
41           41  Assosia88   20  Female        7   
55           55  Phaelap26   25  Female       84   

                       Item Name  Price  
15   Deadline, Voice Of Subtlety   2.89  
18                       Nirvana   4.90  
38       Renewed Skeletal Katana   4.18  
41  Thorn, Satchel of Dark Souls   1.33  
55                    Arcane Gem   3.79

In [22]:
# Female purchase count
female_purchase = len(female_purchase_data)
female_purchase

113

In [23]:
# Average female purchase price
female_mean_price = round(female_purchase_data['Price'].mean(),2)
female_mean_price

3.2

In [24]:
# Total female purchase 
female_total_purchase = round(female_purchase_data['Price'].sum(),2)
female_total_purchase

361.94

In [25]:
# Average purchase total per person by females
female_avg_purchase = round(female_total_purchase/female_players,2)
female_avg_purchase

4.47

In [26]:
# New other/non-disclosed data frame
other_purchase_data = purchase_data.loc[purchase_data['Gender'] == 'Other / Non-Disclosed']
other_purchase_data.head()

Purchase ID            SN  Age                 Gender  Item ID  \
9              9   Chanosian48   35  Other / Non-Disclosed      136   
22            22  Siarithria38   38  Other / Non-Disclosed       24   
82            82    Haerithp41   16  Other / Non-Disclosed      160   
111          111      Sundim98   21  Other / Non-Disclosed       41   
228          228   Jiskirran77   20  Other / Non-Disclosed       80   

                        Item Name  Price  
9    Ghastly Adamantite Protector   3.58  
22                  Warped Fetish   3.81  
82                     Azurewrath   4.40  
111                         Orbit   4.75  
228                     Dreamsong   3.39

In [27]:
# Other purchase count
other_purchase = len(other_purchase_data)
other_purchase

15

In [28]:
# Average other purchase price
other_mean_price = round(other_purchase_data['Price'].mean(),2)
other_mean_price

3.35

In [29]:
# Total other purchase
other_total_purchase = round(other_purchase_data['Price'].sum(),2)
other_total_purchase

50.19

In [30]:
# Average purchase total per person by others
other_avg_purchase = round(other_total_purchase/other_players,2)
other_avg_purchase

4.56

In [31]:
purchase_analysis_gender_df = pd.DataFrame({
    'Gender': ['Female', 'Male', 'Other/Non-Disclosed'],
    'Purchase Count': [female_purchase, male_purchase, other_purchase],
    'Average Purchase Price': [female_mean_price, male_mean_price, other_mean_price],
    'Total Purchase Value': [female_total_purchase, male_total_purchase, other_total_purchase],
    'Avg Total Purchase per Person': [female_avg_purchase, male_avg_purchase, other_avg_purchase]
})
purchase_analysis_gender_df['Average Purchase Price'] = purchase_analysis_gender_df['Average Purchase Price'].map('${:.2f}'.format)
purchase_analysis_gender_df['Total Purchase Value'] = purchase_analysis_gender_df['Total Purchase Value'].map('${:.2f}'.format)
purchase_analysis_gender_df['Avg Total Purchase per Person'] = purchase_analysis_gender_df['Avg Total Purchase per Person'].map('${:.2f}'.format)
purchase_analysis_gender_df = purchase_analysis_gender_df.set_index('Gender')
purchase_analysis_gender_df.index.name = None
purchase_analysis_gender_df

Purchase Count Average Purchase Price  \
Female                          113                  $3.20   
Male                            652                  $3.02   
Other/Non-Disclosed              15                  $3.35   

                    Total Purchase Value Avg Total Purchase per Person  
Female                           $361.94                         $4.47  
Male                            $1967.64                         $4.07  
Other/Non-Disclosed               $50.19                         $4.56

### Result Age Demographics
* Establish bins for ages
* Categorize the existing players using the age bins. Hint: use pd.cut()
* Calculate the numbers and percentages by age group
* Create a summary data frame to hold the results
* Optional: round the percentage column to two decimal points
* Display Age Demographics Table
-----------

In [32]:
# Get min and max age value
purchase_data['Age'].max()

45

In [33]:
purchase_data['Age'].min()

7

In [34]:
age_bins = [0,9, 14, 19, 24, 29, 34, 39, 45]
age_labels = ['<10', '10-14', '15-19', '20-24', '25-29', '30-34', '35-39', '40+']
pd.cut(purchase_data['Age'], age_bins, include_lowest=False, duplicates='drop')

0      (19, 24]
1      (39, 45]
2      (19, 24]
3      (19, 24]
4      (19, 24]
         ...   
775    (19, 24]
776    (19, 24]
777    (19, 24]
778      (0, 9]
779    (19, 24]
Name: Age, Length: 780, dtype: category
Categories (8, interval[int64]): [(0, 9] < (9, 14] < (14, 19] < (19, 24] < (24, 29] < (29, 34] < (34, 39] < (39, 45]]

In [35]:
new_purchase_data = purchase_data
new_purchase_data['Age Range'] = pd.cut(purchase_data['Age'], age_bins, labels=age_labels, include_lowest=False, duplicates='drop')

In [36]:
total_grouped_df = new_purchase_data.groupby(['Age Range'])
total_grouped_df = total_grouped_df.agg({ 'SN': 'nunique' })
total_grouped_df = total_grouped_df.reset_index()

age_demo_df = pd.DataFrame(total_grouped_df)
age_demo_df['Percentage of Players'] = age_demo_df.apply(lambda row: (row[1]/number_of_players)*100, axis=1)
age_demo_df['Percentage of Players'] = age_demo_df['Percentage of Players'].map('{:.2f}%'.format)
age_demo_df = age_demo_df.rename(columns={
    'SN': 'Total Count'
})
age_demo_df = age_demo_df.set_index('Age Range')
age_demo_df.index.name = None
age_demo_df

Total Count Percentage of Players
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

### Result Purchasing Analysis (Age)

* The below each broken into bins of 4 years (i.e. &lt;10, 10-14, 15-19, etc.)
  * Purchase Count
  * Average Purchase Price
  * Total Purchase Value
  * Average Purchase Total per Person by Age Group
----------

In [37]:
age_bins = [0,9, 14, 19, 24, 29, 34, 39, 45]
age_labels = ['<10', '10-14', '15-19', '20-24', '25-29', '30-34', '35-39', '40+']
pd.cut(purchase_data['Age'], age_bins, include_lowest=False, duplicates='drop')
new_purchase_data = purchase_data
new_purchase_data['Age Range'] = pd.cut(purchase_data['Age'], age_bins, labels=age_labels, include_lowest=False, duplicates='drop')

In [38]:
purchas_analysis_grouped = new_purchase_data.groupby(['Age Range'])
purchas_analysis_grouped = purchas_analysis_grouped.agg({
    'Purchase ID': ['count'],
    'Price': ['mean', 'sum'],
    'SN': 'nunique'
})
purchas_analysis_grouped.columns = purchas_analysis_grouped.columns.droplevel(0)
purchas_analysis_age_df = pd.DataFrame(purchas_analysis_grouped)
purchas_analysis_age_df = purchas_analysis_age_df.rename(columns={
    'count':'Purchase Count',
    'mean': 'Average Purchase Price',
    'sum': 'Total Purchase Value',
    'nunique': 'AVG Total per Person By Age'
})
purchas_analysis_age_df['AVG Total per Person By Age'] = purchas_analysis_age_df.apply(lambda row: row['Total Purchase Value']/row['AVG Total per Person By Age'], axis=1)
purchas_analysis_age_df['Average Purchase Price'] = purchas_analysis_age_df['Average Purchase Price'].map('${:.2f}'.format)
purchas_analysis_age_df['Total Purchase Value'] = purchas_analysis_age_df['Total Purchase Value'].map('${:.2f}'.format)
purchas_analysis_age_df['AVG Total per Person By Age'] = purchas_analysis_age_df['AVG Total per Person By Age'].map('${:.2f}'.format)
purchas_analysis_age_df.index.name = None
purchas_analysis_age_df

Purchase Count Average Purchase Price Total Purchase Value  \
<10                23                  $3.35               $77.13   
10-14              28                  $2.96               $82.78   
15-19             136                  $3.04              $412.89   
20-24             365                  $3.05             $1114.06   
25-29             101                  $2.90              $293.00   
30-34              73                  $2.93              $214.00   
35-39              41                  $3.60              $147.67   
40+                13                  $2.94               $38.24   

      AVG Total per Person By Age  
<10                         $4.54  
10-14                       $3.76  
15-19                       $3.86  
20-24                       $4.32  
25-29                       $3.81  
30-34                       $4.12  
35-39                       $4.76  
40+                         $3.19

### Result Top Spenders

* Identify the the top 5 spenders in the game by total purchase value, then list (in a table):
  * SN
  * Purchase Count
  * Average Purchase Price
  * Total Purchase Value
 ----------

In [39]:
top_spenders_grouped = purchase_data.groupby('SN')
top_spenders_grouped = top_spenders_grouped.agg({
    'SN': 'count',
    'Price': ['mean', 'sum'],
})
top_spenders_grouped.columns = top_spenders_grouped.columns.droplevel(0)
top_spenders_grouped
top_spenders_df = pd.DataFrame(top_spenders_grouped)
top_spenders_df = top_spenders_df.sort_values('sum', ascending=False)
top_spenders_df = top_spenders_df.rename(columns={
    'count': 'Purchase Count',
    'mean': 'Average Purchase Price',
    'sum': 'Total Purchase Value'
})
top_spenders_df['Average Purchase Price'] = top_spenders_df['Average Purchase Price'].map('${:.2f}'.format)
top_spenders_df['Total Purchase Value'] = top_spenders_df['Total Purchase Value'].map('${:.2f}'.format)
top_spenders_df.index.name = None
top_spenders_df.head()

Purchase Count Average Purchase Price Total Purchase Value
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

### Result Most Popular Items

* Identify the 5 most popular items by purchase count, then list (in a table):
  * Item ID
  * Item Name
  * Purchase Count
  * Item Price
  * Total Purchase Value
 ---------

In [40]:
most_popular_item_grouped = purchase_data.groupby(['Item ID','Item Name'])
most_popular_item_grouped = most_popular_item_grouped.agg({
    'Item ID': 'count',
    'Price': 'sum'
})
most_popular_item_df = most_popular_item_grouped
most_popular_item_grouped = most_popular_item_grouped.sort_values('Price', ascending=False)
most_popular_item_df['Item Price'] = most_popular_item_df.apply(lambda row: row[1]/row[0], axis=1)
most_popular_item_df = most_popular_item_df.rename(columns={
    'Item ID': 'Purchase Count',
    'Price': 'Total Purchase Value'
}) 
most_profitable_item_df = most_popular_item_df
most_popular_item_df = most_popular_item_df[['Purchase Count', 'Item Price', 'Total Purchase Value']]
most_popular_item_df['Item Price'] = most_popular_item_df['Item Price'].map('${:.2f}'.format)
most_popular_item_df['Total Purchase Value'] = most_popular_item_df['Total Purchase Value'].map('${:.2f}'.format)
most_popular_item_df = most_popular_item_df.sort_values('Purchase Count', ascending=False)
most_popular_item_df.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

### Result Most Profitable Items

* Identify the 5 most profitable items by total purchase value, then list (in a table):
  * Item ID
  * Item Name
  * Purchase Count
  * Item Price
  * Total Purchase Value
 --------

In [41]:
most_profitable_item_df = most_profitable_item_df[['Purchase Count', 'Item Price', 'Total Purchase Value']]
most_profitable_item_df = most_profitable_item_df.sort_values('Total Purchase Value', ascending=False)
most_profitable_item_df['Item Price'] = most_profitable_item_df['Item Price'].map('${:.2f}'.format)
most_profitable_item_df['Total Purchase Value'] = most_profitable_item_df['Total Purchase Value'].map('${:.2f}'.format)
most_profitable_item_df.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

## Observable Trends

### We have a considerable male population
According to the results more than 80% of our players are males, this means that we should focus at least 20% of our efforts with campaigns, items and marketing to this population to satisfy their needs to continue scaling our game. We should remember the 80/20 pareto rule and understand that most of our revenue will come from them and this efforts must include our biggest changes.

### Ages 15 to 24
The data shows that our main player population has an age of 15 to 24 this could mean that they are more willing to spend since probably they still have economic support from their parents. We should consider a marketing campaign focus on highschool and university students with a back to school or student program to grow this market segment.

### Item price for male population
As we can see \\$4.47 is the average total purchase per male, so we can introduce items that range from \\$4.00 to \\$5.00 sepecific for the male gender, like aestehtics items with fire, spikes, and exaggerated names such as 'Critical Sword of Destiny' ,vanity items to change appearance such as 'Height, Muscle body, hairstyles and titles' and other types which focus on game experience ingame quests and more.
